flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeRangeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,596 @@
+        output.replace(rowValue, currAccValue);
+        out.collect(output);
+    }
+
+    private void collectDelete(Collector<RowData> out, RowData rowValue, RowData prevAccValue) {
+        output.setRowKind(RowKind.DELETE);
+        output.replace(rowValue, prevAccValue);
+        out.collect(output);
+    }
+
+    /**
+     * Removes the row matching input from the sortedList. Matching is done based on the sortKeys.
+     * Once matched, the actual row is compared for all ids belonging to the same sortKey. After the
+     * element is removed, a DELETE event is emitted for the removed element, and for all subsequent
+     * elements an UPDATE_BEFORE and UPDATE_AFTER event is emitted based on the previous and newly
+     * aggregated values. Updates are skipped if the previously accumulated value is the same as the
+     * newly accumulated value to save on network bandwidth.
+     *
+     * @param input
+     * @param out
+     * @throws Exception
+     */
+    private void removeFromSortedList(RowData input, Collector<RowData> out) throws Exception {
+        int i;
+        input.setRowKind(RowKind.INSERT);
+        RowData inputSortKey = sortKeySelector.getKey(input);
+        List<Tuple2<RowData, List<Long>>> sortedList = getSortedList();
+
+        for (i = 0; i < sortedList.size(); i++) {
+            RowData curSortKey = sortedList.get(i).f0;
+            List<Long> ids = new ArrayList<>(sortedList.get(i).f1);
+            if (sortKeyEqualiser.equals(curSortKey, inputSortKey)) {
+                setAccumulatorOfPrevRow(sortedList, i - 1);
+                // Accumulate all ids except id which needs to be removed
+                int removeIndex = reAccumulateIdsAndGetRemoveIndex(ids, input);
+                if (removeIndex == -1) {
+                    // TODO: can this ever happen? may be during/after state cleaning?
+                    LOG.info("Could not find id to remove");
+                    return;
+                }
+                emitUpdatesForIds(
+                        ids,
+                        removeIndex,
+                        accMapState.get(curSortKey),
+                        currFunction.getAccumulators(),
+                        RowKind.DELETE,
+                        input,
+                        out);
+                Long deletedId = ids.remove(removeIndex);
+                i = removeIdFromSortedList(sortedList, i, ids, curSortKey);
+                removeFromState(sortedList, deletedId, curSortKey);
+                break;
+            }
+        }
+
+        processRemainingElements(sortedList, i, out);
+    }
+
+    /**
+     * Helper method to re-accumulate the aggregated value for all ids without the id that will be
+     * removed.
+     *
+     * @param ids
+     * @param input
+     * @return the index position of the id that should be removed
+     * @throws Exception
+     */
+    private int reAccumulateIdsAndGetRemoveIndex(List<Long> ids, RowData input) throws Exception {
```

Comment:
```suggestion
    private int reAccumulateIdsAndGetRemoveIndex(List<Long> ids, RowData delRow) throws Exception {
```
easier to understand that the row is about to be deleted / removed.

Reply:
Renamed